core/src/main/java/org/sonarsource/sonarlint/core/log/MessageFormat.java

```diff
@@ -0,0 +1,490 @@
+    if (appendSpecialTypes(o, str)) {
+      return;
+    }
+    if (isMaybeRecursive(o)) {
+      appendPotentiallyRecursiveValue(o, str, dejaVu);
+    } else {
+      tryObjectToString(o, str);
+    }
+  }
+
+  private static boolean appendSpecialTypes(final Object o, final StringBuilder str) {
+    return appendSpecificTypes(str, o) || appendDate(o, str);
+  }
+
+  private static boolean appendSpecificTypes(final StringBuilder stringBuilder, @Nullable final Object obj) {
+    if (obj == null || obj instanceof String) {
+      stringBuilder.append((String) obj);
+    } else if (obj instanceof CharSequence) {
+      stringBuilder.append((CharSequence) obj);
+    } else if (obj instanceof Integer) { // LOG4J2-1437 unbox auto-boxed primitives to avoid calling toString()
+      stringBuilder.append(((Integer) obj).intValue());
+    } else if (obj instanceof Long) {
+      stringBuilder.append(((Long) obj).longValue());
+    } else if (obj instanceof Double) {
+      stringBuilder.append(((Double) obj).doubleValue());
+    } else if (obj instanceof Boolean) {
+      stringBuilder.append(((Boolean) obj).booleanValue());
+    } else if (obj instanceof Character) {
+      stringBuilder.append(((Character) obj).charValue());
+    } else if (obj instanceof Short) {
+      stringBuilder.append(((Short) obj).shortValue());
+    } else if (obj instanceof Float) {
+      stringBuilder.append(((Float) obj).floatValue());
+    } else if (obj instanceof Byte) {
+      stringBuilder.append(((Byte) obj).byteValue());
+    } else {
+      return false;
+    }
+    return true;
+  }
+
+  private static boolean appendDate(final Object o, final StringBuilder str) {
+    if (!(o instanceof Date)) {
+      return false;
+    }
+    final Date date = (Date) o;
+    final SimpleDateFormat format = getSimpleDateFormat();
+    str.append(format.format(date));
+    return true;
+  }
+
+  private static SimpleDateFormat getSimpleDateFormat() {
+    SimpleDateFormat result = threadLocalSimpleDateFormat.get();
+    if (result == null) {
+      result = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ");
+      threadLocalSimpleDateFormat.set(result);
+    }
+    return result;
+  }
+
+  /**
+   * Returns {@code true} if the specified object is an array, a Map or a Collection.
+   */
+  private static boolean isMaybeRecursive(final Object o) {
+    return o.getClass().isArray() || o instanceof Map || o instanceof Collection;
+  }
+
+  private static void appendPotentiallyRecursiveValue(final Object o, final StringBuilder str,
+    final Set<String> dejaVu) {
+    final Class<?> oClass = o.getClass();
+    if (oClass.isArray()) {
+      appendArray(o, str, dejaVu, oClass);
+    } else if (o instanceof Map) {
+      appendMap(o, str, dejaVu);
+    } else if (o instanceof Collection) {
+      appendCollection(o, str, dejaVu);
+    }
+  }
+
+  private static void appendArray(final Object o, final StringBuilder str, @Nullable Set<String> dejaVu, final Class<?> oClass) {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Refactor this method to reduce its Cognitive Complexity from 22 to the 15 allowed. ([squid:S3776](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=165&issues=AWM_D9R15HoPdzsInowT&open=AWM_D9R15HoPdzsInowT)

[ISSUE_KEY]: AWM_D9R15HoPdzsInowT